executor/parallel_apply.go

```diff
@@ -0,0 +1,389 @@
+}
+
+// ParallelNestedLoopApplyExec is the executor for apply.
+type ParallelNestedLoopApplyExec struct {
+	baseExecutor
+
+	// outer-side fields
+	cursor        int
+	outerExec     Executor
+	outerFilter   expression.CNFExprs
+	outerList     *chunk.List
+	outerRowMutex sync.Mutex
+	outer         bool
+
+	// inner-side fields
+	// use slices since the inner side is paralleled
+	corCols       [][]*expression.CorrelatedColumn
+	innerFilter   []expression.CNFExprs
+	innerExecs    []Executor
+	innerList     []*chunk.List
+	innerChunk    []*chunk.Chunk
+	innerSelected [][]bool
+	innerIter     []chunk.Iterator
+	outerRow      []*chunk.Row
+	hasMatch      []bool
+	hasNull       []bool
+	joiners       []joiner
+
+	// fields about concurrency control
+	concurrency int
+	started     bool
+	freeChkCh   chan *chunk.Chunk
+	resultChkCh chan result
+	outerRowCh  chan outerRow
+	exit        chan struct{}
+	workerWg    sync.WaitGroup
+	notifyWg    sync.WaitGroup
+
+	// fields about cache
+	cache              *applyCache
+	useCache           bool
+	cacheHitCounter    int64
+	cacheAccessCounter int64
+	cacheLock          sync.RWMutex
+
+	memTracker *memory.Tracker // track memory usage.
+}
+
+// Open implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Open(ctx context.Context) error {
+	err := e.outerExec.Open(ctx)
+	if err != nil {
+		return err
+	}
+	e.memTracker = memory.NewTracker(e.id, -1)
+	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
+
+	e.outerList = chunk.NewList(retTypes(e.outerExec), e.initCap, e.maxChunkSize)
+	e.outerList.GetMemTracker().SetLabel(stringutil.StringerStr("outerList"))
+	e.outerList.GetMemTracker().AttachTo(e.memTracker)
+
+	e.innerList = make([]*chunk.List, e.concurrency)
+	e.innerChunk = make([]*chunk.Chunk, e.concurrency)
+	e.innerSelected = make([][]bool, e.concurrency)
+	e.innerIter = make([]chunk.Iterator, e.concurrency)
+	e.outerRow = make([]*chunk.Row, e.concurrency)
+	e.hasMatch = make([]bool, e.concurrency)
+	e.hasNull = make([]bool, e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		e.innerChunk[i] = newFirstChunk(e.innerExecs[i])
+		e.innerList[i] = chunk.NewList(retTypes(e.innerExecs[i]), e.initCap, e.maxChunkSize)
+		e.innerList[i].GetMemTracker().SetLabel(innerListLabel)
+		e.innerList[i].GetMemTracker().AttachTo(e.memTracker)
+	}
+
+	e.freeChkCh = make(chan *chunk.Chunk, e.concurrency)
+	e.resultChkCh = make(chan result, e.concurrency+1) // innerWorkers + outerWorker
+	e.outerRowCh = make(chan outerRow)
+	e.exit = make(chan struct{})
+	for i := 0; i < e.concurrency; i++ {
+		e.freeChkCh <- newFirstChunk(e)
+	}
+
+	if e.useCache {
+		if e.cache, err = newApplyCache(e.ctx); err != nil {
+			return err
+		}
+		e.cache.GetMemTracker().AttachTo(e.memTracker)
+	}
+	return nil
+}
+
+// Next implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
+	if !e.started {
```

Comment:
Is it possible that there is a parent operator of `apply` which could call the `apply.Next` concurrently, then may the start not only once `outerWorker` and `innerWorker`.

Reply:
You are right... I update this line and an atomic operation `CAS` is used now.
And I find `HashJoin` has the same code `if !e.prepare`, but I don't know why there is no problem with it so far...